persist: extract an Applier from Machine

danhhz opened the pull request:
Applier exists mainly to allow us to very narrowly bound the surface area that directly interacts with state. This sets us up to wrap State in an Arc<tokio::sync::Mutex<_>> and share them between machines (one copy of state per shard per process).

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

The intent of this is to be a pure refactor, with no behavior changes.

Definitely go commit-by-commit:
- The first commit is pure code movement.
- The second commit intends to shuffle some things around in prep for the third commit, but critically having no meaningful impact. As close to pure refactor as reasonably possible. If we decide it's desirable, I structured things such that split this into 2 PRs here.
- The third commit is a minor error cleanup that would have made a unit test failure easier to debug.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

danhhz commented:
Okay, this still needs some metrics, but I've been staring at it for too long and it's time to get other eyes on it. @pH14 and @bkirwi please take a look! In particular, I'd love to hash out the WIP questions I left in there

pH14 commented on src/persist-client/src/cache.rs:
```diff
@@ -192,4 +370,168 @@ mod tests {
+        assert_eq!(states.initialized_count().await, 2);
+        assert_eq!(states.strong_count().await, 2);
+        drop(s1_state1);
+        assert_eq!(states.strong_count().await, 1);
+    }
```
love this test!

pH14 commented on src/persist-client/src/cache.rs:
```diff
@@ -120,10 +133,175 @@ impl PersistClientCache {
     }
 }
 
+trait DynState: Debug + Send + Sync {
```
is there a tl;dr on why we need `DynState`? this is far past my ability to reason about rust-isms on paper, but I'm a little surprised about this vs making `StateCache` generic over KVTD

danhhz replied:
I guess we don't really do this in practice (yet?) but otherwise, we'd have to actually carry the K, V, T, D, types all the way up to PersistClient, which really doesn't seem right to me

pH14 commented on src/persist-client/src/internal/apply.rs:
```diff
@@ -0,0 +1,436 @@
+        &self,
+        as_of: &Antichain<T>,
+    ) -> Result<Result<Vec<HollowBatch<T>>, Upper<T>>, Since<T>> {
+        // WIP safe to serialize this on the mutex?
```
what's the definition of safety for the WIP ones here? safe in terms of correctness? or in terms of performance characteristics

pH14 replied:
I can't think of why `snapshot` would be unsafe from either perspective, we get all of the batches from a single version of state in one go, and it should be quick to determine those batches

danhhz replied:
oh, sorry. performance and availability. this e.g. would have to serialize with the commands wanting to lock the mutex

pH14 commented on src/persist-client/src/internal/apply.rs:
```diff
@@ -0,0 +1,436 @@
+        as_of: &Antichain<T>,
+    ) -> Result<Result<Vec<HollowBatch<T>>, Upper<T>>, Since<T>> {
+        // WIP safe to serialize this on the mutex?
+        self.state.lock().await.snapshot(as_of)
```
that said, at the risk of more complexity, I wouldn't hate using a `RwLock`. It seems like snapshots and listens could use read locks, and `apply_unbatched_cmd`, `fetch_and_update_state`, etc could hold write locks and minimize contention. not sure how easy it is to instrument, but could be interesting to see how much time we spend waiting on the lock within each operation

danhhz replied:
I'm happy to use a RwLock, is it even more complexity? I'm not sure if it'll super help a ton though, because RwLocks shine when there are lots of reads and few writes but here writes are both far more common and in the critical section across network calls, while reads probably won't be super contend-y

metrics around this lock are definitely something we'll add before merging, I just ran out of stream on friday

pH14 commented on src/persist-client/src/internal/apply.rs:
```diff
@@ -0,0 +1,436 @@
+
+    pub async fn next_listen_batch(&self, frontier: &Antichain<T>) -> Option<HollowBatch<T>> {
+        // WIP safe to serialize this on the mutex?
+        self.state.lock().await.next_listen_batch(frontier)
```
it seems like it's also possible that `cached_state` contains a full copy of the latest `State<K, V, T, D>`. that means we always keep 2 copies of State per shard per process, but would mean we could run all reads without any locking at all. offhand that doesn't seem like the right tradeoff, but I did want to make sure it was stated out loud somewhere :)

danhhz replied:
something I probably didn't explain well is that there is one State per shard per process, but CachedState is the same one per Machine/Handle model as before. so if it was a full copy, we'd be back where we started

however, one thing that this does suggest to me though is two shared copies, one of which looks like I have it, and the other would only get the lock very briefly at the end of an apply_unbatched_cmd or fetch_and_update_state and be used only for snapshot/listen/etc and other read things. this could also maybe even replace CachedState. downsides are:
- we'd have to reason about two mutexes (potential deadlocks, potential additional latency if something surprising happens), I wonder if we could use something like AtomicPtr?
- 2 copies of state instead of one (seems fine IMO)
- complexity (unfortunate but probably fine)
- the second copy probably involves a lot of cloning (bummer but certainly better than stressing about the contention we're stressing about in the WIP question I have)

pH14 replied:
ooh I see, didn't grok that CachedState is still per handle. the two shared copies is pretty interesting... it does seem very liberating to not worry about the impact of reads, but I'm not sure how that compares to reasoning about two mutexes 🤔

danhhz replied:
I think I've gradually talked myself into the two mutex thing as this has sat, but we obviously want to make that change extremely carefully.

I'm also starting to worry about messing up rebases here, so as a result, I think I'd like to split the `state: State` -> `Arc<Mutex<State>>` bit into its own followup PR and leave everything else here to merge first. thoughts?

pH14 replied:
sounds good, though would we need to worry about shipping just this PR in a release?

danhhz replied:
in the "scary change" sense? commits 1+2+4 are meant to be as close to pure code movement as possible. the only real exception to that is the introduction of CachedState stuff, but I could easily pull that out too

pH14 replied:
yes to the "scary change" sense, though I think I misunderstood the original question. I'm comfortable with commits 1+2+4 shipping first

pH14 commented on src/storage/src/sink/healthcheck.rs:
```diff
@@ -480,17 +480,19 @@ mod tests {
+                    (Result<SourceData, String>, Result<(), String>),
+                    Timestamp,
+                    i64,
+                )| { v.unwrap().0.unwrap() },
```
misc: imagine showing someone unfamiliar with Rust these 5 lines of code and informing them that this is one of the most beloved programming languages

pH14 reviewed (commented):
Lookin' good! The code movement makes this a more well contained change than I would have expected, reads nicely. I'd love to see the release qualification tests pointed at this before merging, and I have a question about the meaning of safety for the read ops you flagged before digging into them further

danhhz reviewed (commented):
> I'd love to see the release qualification tests pointed at this before merging

Oh meant to say this myself, on board 100%. I had actually been roughly planning on handing it to philip and telling him to do his worst, in case he could come up with things that were even more interesting than nightly+release qual. He's out this week, so one thing we could do, if we liked, is merge the first two commits (should a no-op, roughly, but contain most of the stuff that would make a rebase scary) and then leave the last two for when he's back.

> The code movement makes this a more well contained change than I would have expected

Thanks! This is the *check notes* fourth attempt at this PR and the first one I felt was mergeable.

danhhz commented:
updated this to be just the refactor bits, so it's ready to go! I also rebased in main, which was pretty substantial. this both confirms my intuition that I'd like to get this PR landed sooner rather than later (the 3rd commit that we're pulling out shouldn't be quite as tricky to rebase) and also means it's probably best to just review this from scratch

danhhz commented:
@pH14 @bkirwi friendly ping on this! I'd love to get it merged before it rots again

pH14 commented on src/persist-client/src/cli/inspect.rs:
```diff
@@ -778,8 +778,14 @@ struct T;
 
-pub(crate) static KVTD_CODECS: Mutex<(String, String, String, String)> =
-    Mutex::new((String::new(), String::new(), String::new(), String::new()));
+pub(crate) static KVTD_CODECS: Mutex<(String, String, String, String, Option<&'static str>)> =
```
nit: feels like this deserves a newtype or some simple struct to explain the semantics of the last element a bit more

danhhz replied:
done!

pH14 reviewed (approved):
Looks like nearly all non-scary code movement + the movement makes sense = LGTM 🚢

bkirwi commented on src/persist-client/src/error.rs:
```diff
@@ -182,9 +182,15 @@ impl<T> Determinacy for InvalidUsage<T> {
+    ///
+    /// The last element in the tuple is Some when the name of the codecs match,
+    /// but the concrete types don't: e.g. mz_repr::Timestamp and u64.
+    pub(crate) requested: (String, String, String, String, Option<&'static str>),
```
The new degree of freedom doesn't seem to be used here... I assume that's expected and setup for future changes, but FYI in case it is not!

danhhz replied:
yup, known! it got used in the commit that's getting pulled out into a second PR

bkirwi reviewed (approved):
All seems sensible!

danhhz commented:
TFTRs!
